src/providers/pkcs11/mod.rs

```diff
@@ -479,27 +487,75 @@ impl ProviderBuilder {
                 Error::new(ErrorKind::InvalidData, "error initializing PKCS 11 context")
             })?;
 
-        let slot_number = match self.slot_number {
-            Some(i) => {
-                let slot = Slot::try_from(i).or_else(|_| {
+        let slots = backend.get_slots_with_initialized_token().map_err(|e| {
+            format_error!(
+                "Failed retrieving a valid slot with an initialized token",
+                e
+            );
+            Error::new(
+                ErrorKind::InvalidData,
+                "Failed retrieving a valid slot with an initialized token",
+            )
+        })?;
+        let slot_number = match (self.serial_number, self.slot_number) {
+            (Some(serial_number), _given_slot) => {
+                let mut slot = None;
+                for current_slot in slots {
+                    let current_token = backend.get_token_info(current_slot).map_err(|e| {
+                        format_error!("Failed retrieving token info", e);
+                        Error::new(ErrorKind::InvalidData, "Failed retrieving token info")
+                    })?;
+                    let sn =
+                        String::from_utf8(current_token.serialNumber.to_vec()).map_err(|e| {
+                            format_error!("Failed retrieving token serial number", e);
+                            Error::new(
+                                ErrorKind::InvalidData,
+                                "Failed retrieving token serial number",
+                            )
```

Comment:
```suggestion
                            format_error!("Failed parsing token serial number", e);
                            Error::new(
                                ErrorKind::InvalidData,
                                "Failed parsing token serial number",
                            )
```

Though this will go away once combined with the `cryptoki` update.